MetaDataset.py

```diff
@@ -885,23 +886,112 @@ def init_seq_order(self, epoch=None, seq_list=None):
     """
     assert seq_list is None, "seq_list not supported for %s" % self.__class__
     need_reinit = self.epoch is None or self.epoch != epoch
-    super(CombinedDataset, self).init_seq_order(epoch=epoch, seq_list=seq_list)
+    num_seqs_saved = self._num_seqs
+    super(CombinedDataset, self).init_seq_order(epoch=epoch, seq_list=seq_list)  # resets self._num_seqs
     self.rnd.seed(epoch or 1)
 
     if not need_reinit:
+      self._num_seqs = num_seqs_saved
       return False
 
-    # First init sequence order for sub-datasets as usual to get a list of available sequences. This way sorting and
+    # First init sequence order for sub-datasets as usual to get a list of available sequences. This way, sorting and
     # partition epoch of the individual sub-datasets is still supported. Later we will call init_seq_order again with a
     # sequence list to e.g. apply joint sorting or partition epoch of all sequences.
     for dataset in self.datasets.values():
       dataset.init_seq_order(epoch=epoch)
 
-    self.dataset_sorted_seq_idx_list = []  # We will fill this as we go
-    self.used_num_seqs_per_subset = [0] * len(self.datasets)
+    try:
+      total_num_seqs = sum([self.datasets[k].num_seqs for k in sorted(self.datasets.keys())])
+      # Whether we can get num_seqs from all sub-datasets should not change between epochs, therefore
+      # (self._can_use_num_seqs is False) should never happen here. However, to be sure we keep it False in such a case
+      # to not change behaviour between epochs.
+      if self._can_use_num_seqs is None:
+        self._can_use_num_seqs = True
+    except Exception:
+      # Here we cannot work around changing behaviour (see comment above).
+      assert not self._can_use_num_seqs, (
+        "Failed to get num_seqs from all sub-datasets although this was possible in a previous epoch.")
```

Comment:
Why is this a problem?

Reply:
Ok, I will allow that then. But note that sequence orderings can mean slightly different things, depending on num_seqs being available. E.g "random_dataset" (which I plan to add back in the next commit for the known num_seqs case). In one case you sample from the "true distribution", in the other you rely on estimated_num_seqs.
Also when using ordering not supported if num_seqs is not known, this will cause an error later anyway.

Reply:
I don't quite understand. If you want to check `seq_ordering`, you can check for `seq_ordering`.
But here you are checking `_can_use_num_seqs`. `_can_use_num_seqs ` would be some value from a previous epoch (or `None`). Why does it matter what it was in the previous epoch? That's what I don't understand.

Reply:
There are two in principle very different implementations creating the sequence order. Either globally, if num_seqs is known, or just for the next few sequences, in case num_seqs is unknown. The question is whether we want to allow switching between the two from one epoch to the other. For this we would compare to the previous value of `_can_use_num_seqs`. Switching is only possible if the sequence ordering is implemented for both cases, which is not the case in general (often even not possible). So we can either prohibit it in all cases, or allow it and later `_expand_dataset_sec_idxs`might might fail because of an unsupported ordering.

Reply:
Because it might not be intended that when you set "random_dataset" as a sequence ordering in the config, that this means different things for different epochs. But ok, I will just remove the checks.

Reply:
I still don't understand. Why is it not ok if it changes for different epochs, but ok otherwise? I would say, either it is ok that the setting can mean different things, or it is not ok.

I would rather say that it should not be solved like that in any case. There should be distinct settings in the config, for different behaviors. A setting should mean one behavior. E.g. if you have some global random order if `num_seqs` is known, call it `global_random_dataset_order` or so, and otherwise sth else. And then if you use `global_random_dataset_order` but `num_seqs` is not available, you would throw an exception.